### DISCUSSION TITLE
Where is the user-role relationship stored?

### DISCUSSION THREAD

**Discussion by zhanilyasov** [spatie/laravel-permission#2134] [Q&A]
I opened the database and did not find a table with relationships there. Why are they not in the database and where are they stored?

**Comment by erikn69** [comment]
`model_has_roles`, `model_has_permissions`
https://github.com/spatie/laravel-permission/blob/main/database/migrations/create_permission_tables.php.stub

**Reply by erikn69** [reply]
`model_has_roles`, `model_has_permissions` are [Polymorphic-Relations](https://laravel.com/docs/9.x/eloquent-relationships#many-to-many-polymorphic-relations)

>For user model the model_type is "App\Models\User" and the model_id attribute in model_has_roles seem to be a foreign key to the id in the users table.
>Is this about correct, @erikn69?

Yes, that is correct, also as polymorphic, you could have other models like "App\Models\Client", "App\Models\Admin" and others

>What I wonder about is what is the best way to populate the roles for new users. Should I have one blade for user registration and one for role assignment or mix it?

It depends on you, works the same to me

>Should I then create a new model for the model_has_roles table?

No, laravel and this package handles that with relations, read ([Polymorphic-Relations](https://laravel.com/docs/9.x/eloquent-relationships#many-to-many-polymorphic-relations), [Assigning  Permissions](https://spatie.be/docs/laravel-permission/v5/basic-usage/direct-permissions), [Assigning Roles](https://spatie.be/docs/laravel-permission/v5/basic-usage/role-permissions#content-assigning-roles))


**Mark this as answered** if you don't have any more questions

**Reply by acharseth** [reply]
In app/Actions/Fortify/CreateNewUser I have the following create function:

    public function create(array $input)
    {
            Validator::make($input, [
                'name' => ['required', 'string', 'max:255'],
                'email' => ['required', 'string', 'email', 'max:255', 'unique:users'],
                'password' => $this->passwordRules(),
                'terms' => Jetstream::hasTermsAndPrivacyPolicyFeature() ? ['accepted', 'required'] : '',
            ])->validate();
          
          $newUser = new User;
          $newUser -> create([
                             'name' => $input['name'],
                             'email' => $input['email'],
                             'password' => Hash::make($input['password']),
                         ]);
          
          $newUser->assignRole($input['roles']);
          return $newUser;
	}

This runs fine when user is created but I don't see the roles added in the model_has_roles table. Aren't the roles supposed to be populated here by the `$newUser->assignRole($input['roles']);`,  @erikn69?

PS: For debugging purposes I have tried including this statement in addition just to be sure the problem isn't with the parsing of the input: `$newUser->assignRole(Role::findById(2));`, but this does't cause any changes in the model_has_roles table either.

**Reply by erikn69** [reply]
>Aren't the roles supposed to be populated here by the $newUser->assignRole($input['roles']);, @erikn69?

Yes

Try this 
```php
$newUser = User::create([
     'name' => $input['name'],
     'email' => $input['email'],
     'password' => Hash::make($input['password']),
 ]);

$newUser->assignRole($input['roles']);
return $newUser;

```
Also take a look on `$input['roles']`

**Reply by acharseth** [reply]
Thanks a lot, that worked! (Sorry for newbie question).

**Comment by acharseth** [comment]
@erikn69 indirectly answered this by pointing you to the migration file showing the table structure.
So how I interpret these tables is that user roles are stored in model_has_roles table where user is a model (meaning that you could assign roles to other models than user). So the database model is something like this:
	                                              
	roles -< model_has_roles >-(model_type)- model

( «-<» is a text based «crow's foot» ER model notation denoting a 1:N relationship. The parantheses in the second relationship denotes that the relationship is of a type (with attributt model_type concretely in this case) identifying the specific model in question).

For user model the model_type is "App\Models\User" and the model_id attribute in model_has_roles seem to be a foreign key to the id in the users table.

Graphical with user as model:
![UserRoles](https://user-images.githubusercontent.com/16928675/185797165-8c7e0577-c873-40b6-ba73-8b31943e323a.jpg)


Is this about correct, @erikn69?

What I wonder about is what is the best way to populate the roles for new users. Should I have one blade for user registration and one for role assignment or mix it? (Probably one for each). Should I then create a new model for the model_has_roles table?
